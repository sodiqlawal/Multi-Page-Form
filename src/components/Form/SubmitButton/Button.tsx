import React from "react";
import classNames from "classnames";
import CircularProgress from "@material-ui/core/CircularProgress";

export interface TButtonProps
  extends React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  > {
  content: string;
  isLoading?: boolean;
}
const Button = (props: TButtonProps) => {
  const { isLoading, content, ...restProps } = props;

  return (
    // eslint-disable-next-line react/button-has-type
    <button
      {...restProps}
      disabled={!!isLoading}
      className={classNames("button", props.className, {
        isSubmitting: !!isLoading,
      })}
      data-testid="button"
    >
      {props.isLoading ? (
        <CircularProgress size={15} thickness={6} color="inherit" />
      ) : (
        <span data-testid="content">{content}</span>
      )}
    </button>
  );
};

export default Button;
