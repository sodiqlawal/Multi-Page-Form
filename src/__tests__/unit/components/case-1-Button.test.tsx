import React from "react";
import { render, fireEvent } from "@testing-library/react";

import Button, { TButtonProps } from "components/Form/SubmitButton/Button";
import ReactDOM from "react-dom";

describe("The <Button /> component", () => {
  const defaultButtonProps: TButtonProps = {
    isLoading: false,
    content: "click",
    onClick: jest.fn(),
  };
  const setupButton = (props = defaultButtonProps) =>
    render(<Button {...props} />);

  it("Should render the button and the content user will see", () => {
    // using react dom
    const container = document.createElement("div");
    ReactDOM.render(<Button {...defaultButtonProps} />, container);
    expect(container.querySelector("button")).not.toBeNull();

    const { asFragment, getByText } = setupButton();
    expect(asFragment()).toMatchSnapshot();
    expect(getByText(defaultButtonProps.content)).toBeInTheDocument();
  });

  it("Should call the click handler when provided", () => {
    const { getByTestId } = setupButton();
    const button = getByTestId("button");
    fireEvent.click(button);
    expect(defaultButtonProps.onClick).toHaveBeenCalled();
  });

  it("Should remove content text when loading", () => {
    const { queryByTestId } = setupButton({
      ...defaultButtonProps,
      isLoading: true,
    });

    expect(queryByTestId("content")).toBeNull();
  });
});
