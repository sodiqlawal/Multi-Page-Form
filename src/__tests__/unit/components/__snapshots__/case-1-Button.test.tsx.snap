// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <Button /> component Should render the button and the content user will see 1`] = `
<DocumentFragment>
  <button
    class="button"
    data-testid="button"
  >
    <span
      data-testid="content"
    >
      click
    </span>
  </button>
</DocumentFragment>
`;
